graph_worker.cpp

```diff
@@ -0,0 +1,78 @@
+#include "include/graph_worker.h"
+#include "include/graph.h"
+#include <buffer_tree.h>
+
+#include <fstream>
+#include <string>
+
+bool GraphWorker::shutdown = false;
+int GraphWorker::num_groups = 1;
+int GraphWorker::group_size = 1;
+const char *GraphWorker::config_file = "graph_worker.conf";
+GraphWorker **GraphWorker::workers;
+
+void GraphWorker::startWorkers(Graph *_graph, BufferTree *_bf) {
+	std::string line;
+	std::ifstream conf(config_file);
+	if (conf.is_open()) {
+		while(getline(conf, line)) {
+			if(line.substr(0, line.find('=')) == "num_groups") {
+				num_groups = std::stoi(line.substr(line.find('=')+1));
+				printf("Number of groups = %i\n", num_groups);
+			}
+			if(line.substr(0, line.find('=')) == "group_size") {
+				group_size = std::stoi(line.substr(line.find('=')+1));
+				printf("Size of groups = %i\n", group_size);
+			}
+		}
+			
+	} else {
+		printf("WARNING: Could not open thread configuration file!\n");
+	}
+	workers = (GraphWorker **) calloc(num_groups, sizeof(GraphWorker *));
+	for (int i = 0; i < num_groups; i++) {
+		workers[i] = new GraphWorker(i, _graph, _bf);
+	}
+
+	shutdown = false;
+}
+
+void GraphWorker::stopWorkers() {
+	shutdown = true;
+	for (int i = 0; i < num_groups; i++) {
+		delete workers[i];
+	}
+	delete workers;
+}
+
+GraphWorker::GraphWorker(int _id, Graph *_graph, BufferTree *_bf) :
+  id(_id), graph(_graph), bf(_bf), thr(startWorker, this) {
+}
+
+GraphWorker::~GraphWorker() {
+	thr.join();
```

Comment:
have you considered detaching the thread instead of waiting to join?

Reply:
No. We want to wait on the thread finishing. The shutdown message is not an immediate stop rather it just tells the thread to exit when there is no more work to do.

Reply:
Ah ok, and we're doing that implicitly with the delete call. Maybe document this somewhere.